frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -119,6 +120,12 @@ public synchronized void removeContainerRequest(ResourceDescriptor resource, Lis
   public synchronized SelectionResult select(
       ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType)
       throws NotAvailableException {
+    return select(requestResource, requestNodeLabel, requestNodeGpuType, 1);
+  }
+
+  public synchronized SelectionResult select(
```

Comment:
>select [](start = 38, length = 6)

looks like "select" should only be called after candidateNodes are ready. Can you add a flag in SelectionManager to indicate that candidateNodes are ready? And everytime when calling select you should check if the flag is set. throw exception if the flag is not set

Reply:
outside  the SelectionManager, each time call ResourceAllocation, applicationmater will get the lastest NodeReport and update the candidateNode.  so the above situation will not happened. 

Even if this situation happened,  the selection will return 0 means no candidate node selected, the AM can do retry or relax.  I think this is "equal" with exception to say candidatenode not ready. 

The selection scenario was covered in UT. 


---
In reply to: [166909484](https://github.com/Microsoft/pai/pull/318#discussion_r166909484) [](ancestors = 166909484)